File: backend/Api/Handler.cs
using System.Text.Json;
using System.Threading.Tasks;

namespace Api
{
    public interface IHandler
    {
        Task<string> Run(string requestString);
    }

    public abstract class BaseHandler<TRequest, TResponse> : IHandler
    {
        protected abstract Task<TResponse> Handle(TRequest request);

        public async Task<string> Run(string requestString)
        {
            var request = JsonSerializer.Deserialize<TRequest>(requestString);

            var response = await Handle(request);

            return JsonSerializer.Serialize(response);
        }
    }
}

File: packages/backend/src/Api/Handler.ts
import { Serializer } from "@tsukiy0/tscore";

export abstract class Handler<TRequest, TResponse> {
  constructor(
    private readonly requestSerializer: Serializer<TRequest, unknown>,
    private readonly responseSerializer: Serializer<TResponse, unknown>,
  ) {}

  abstract handle(request: TRequest): Promise<TResponse>;

  async run(request: unknown): Promise<unknown> {
    return this.responseSerializer.serialize(
      await this.handle(this.requestSerializer.deserialize(request)),
    );
  }
}
